### DISCUSSION TITLE
Guide for Automatic Boot of a Mac using this Utility (WIP)

### DISCUSSION THREAD

**Discussion by 0nelight** [0nelight/macOS-Disable-RAM-Areas#3] [General]
### THIS IS WORK-IN-PROGRESS - PLEASE FOLLOW DISCUSSION TO SEE WHAT IS WORKING TILL NOW

In case you successfully tested a manual startup of a Mac with the this EFI-Utility you may be intererested to automate booting this on every boot.

What you ask for is an so called "NSH-Script" - it is a Textfile which EFI-Shell commands in it that gets interpreted line-by-line by an EFI-Shell.

Here is how you do it:

1. **Make sure you have a working EFI-Shell which you can access through rEFInd**
An EFI-Shell is a small program like the Windows cmd or powershell but for the EFI-Environment. You can download one [here.](https://github.com/tianocore/edk2/blob/UDK2018/EdkShellBinPkg/FullShell/X64/Shell_Full.efi) 

2. **Create an Textfile with an .nsh extension**
Regarding the possible commands - you can refer to ([Shell Command Reference Manual](https://sourceforge.net/projects/efidevkit/files/Documents/ShellCommandManual.pdf/download)).
```
echo Running disable-ram-area.efi with Parameters
stall 1000000
fs7:\EFI\tools\disable-ram-area.efi 0x190000000 0x21000000 3000
echo Starting MacOs
stall 1000000
fs5:\5668A9CD-407D-35EB-87EB-59CD393D65B1\System\Library\CoreServices\boot.efi
```

3. **Create menuentry in refind.conf to be able to call this NSH-File**
Regarding creating [Manual Boot Stanzas](https://www.rodsbooks.com/refind/configfile.html#stanzas).
in loader you put the path for your downloaded efi-shell
in options you put the path for your NSH script
```
menuentry "Boot MacOs with defective RAM-Area" {
    loader /EFI/tools/Shell_Full.efi
    options "dummyarg fs7:\EFI\tools\launch_macoswithdefectiveram.nsh exit"
}
```
4. **Set this as default in rEFInd**
With the ["default_selection"](https://www.rodsbooks.com/refind/configfile.html#adjusting) Token.

**Comment by daniel-nextgti** [comment]
Hi @0nelight 

I`ve tried the method above. The .nsh-file is read but automatically kicks me out to the shell terminal and neither commands on the file are read by the shell.

I`ve recorded what happened in Slo-Mo, just so it could be read in time. Here are some prints:
<img width="872" alt="Screenshot 2024-09-11 at 11 06 20" src="https://github.com/user-attachments/assets/b78b102a-45ed-48da-99b8-82582dd7d529">
<img width="989" alt="Screenshot 2024-09-11 at 11 06 27" src="https://github.com/user-attachments/assets/e23a924c-6876-45b8-83b1-d9c5572b370a">
<img width="847" alt="Screenshot 2024-09-11 at 11 06 35" src="https://github.com/user-attachments/assets/df73f8a1-74f8-466b-8513-946782b07461">
<img width="850" alt="Screenshot 2024-09-11 at 11 06 42" src="https://github.com/user-attachments/assets/df22b206-4f14-447c-b429-9d3e8ce1c3f2">



I`ve ajusted the script to the correct paths:
<img width="1069" alt="Screenshot 2024-09-11 at 11 08 08" src="https://github.com/user-attachments/assets/19aa2894-b4ae-487a-a973-a25e4088aed3">


And, here`s the rEFInd.config menuentry:
<img width="1037" alt="Screenshot 2024-09-11 at 11 08 46" src="https://github.com/user-attachments/assets/e3258110-2264-470e-90d8-c97a92b6cc1b">


Any ideas why the script didn`t worked?

Thanks again,
Dan

**Reply by MeTalOneOEight** [reply]
At least the Booting MacOS path in the .nsh looks wrong.
It's pointing to an installer.
Compare it to the rEFInd entry.
Better check all paths.

**Comment by 0nelight** [comment]
So the following nsh-file works for me - but only with the compiled version with the hard-coded ram-areas.
When trying with the command line arguments version of disable-ram-area.efi - then I get a hang immediately after entering it.

Here the Content of **launch_macoswithdefectiveram.nsh**:

```
echo Running disable-ram-area.efi with Parameters
stall 1000000
fs7:\EFI\tools\disable-ram-area.efi 0x190000000 0x21000000 3000
echo Starting MacOs
stall 1000000
fs5:\5668A9CD-407D-35EB-87EB-59CD393D65B1\System\Library\CoreServices\boot.efi
```

here the menuentry at the End of **refind.conf**:
```
menuentry "Boot MacOs with defective RAM-Area" {
    loader /EFI/tools/Shell_Full.efi
    options "dummyarg fs7:\EFI\tools\launch_macoswithdefectiveram.nsh exit"
}
```

Please note the first argument under options "dummyarg" you could write anything here - it is necessary for the way the EFI-Shell takes in the arguments. It throws aways the first one.

Also note we have to use backslashes in the options-line.

Please note that fs0: fs5: and so on could be different in your system. Best to navigate through them interactively to get an overview. Simply enter Shell_Full.efi via rEFInd Boot Manager and type in "help" - you will see all the commands.

### Edit 1:
Added Command Line Options within the NSH File

**Reply by daniel-nextgti** [reply]
Hello @0nelight 

With this code, the .nsh is called using the options line, but with the .efi, i need to input the RAM range on the options line, too.
Where should i put the correct range while using the options to call the .nsh file?

I got confused how to use the .efi version with range args while using the .nsh

Thanks again for all your effort!
Dan

**Reply by 0nelight** [reply]
You are right! I corrected my previous post.

In refind.conf the **options** line is the path to the NSH File. 

In the NSH File you put the path to the disable-ram-area.efi and its parameters into the same line - each after the other (see post for example).

**Reply by daniel-nextgti** [reply]
@0nelight Thanks for updating the code above!

Unfortunately, when i use the lastest .efi file and the code above, i got an error with the parsed values - Even with the correct StartAddr/EndAddr, the script reads a different StartAddr, and uses the StartAddr that i`ve inputed as the EndAddr. 

Please, take a look at the picture below: 
![Screenshot 2024-09-17 at 15 43 16](https://github.com/user-attachments/assets/514abb6b-c644-4426-9ab5-da9561505f37)

Any ideas why the script is parsing a different start value?

Thanks,
Dan

**Reply by 0nelight** [reply]
This is connected to how the Efi-Shell-App provides the command line parameters to the disable-ram-areas.efi-App.

I will think about a solution for this. Maybe I have to change the Code.

**Reply by 0nelight** [reply]
So the refind-boot-manager only passes the three arguments to the efi-app while the efi-shell passes an additional parameter at the beginning - the name of the app. 

so my efi-app gets 4 parameters instead of 3 which causes the error.

I will fix this asap so that the app counts how many parameters are provided and in case they are 4 the first one will be skipped :)

**Reply by 0nelight** [reply]
[Updated Release](https://github.com/0nelight/macOS-Disable-RAM-Areas/releases/tag/v0.2.2-beta) - Please test and tell me your results!

**Reply by daniel-nextgti** [reply]
Hello @0nelight !

Sorry for keeping you waiting for so long!
I have to thank you once again for the amazing work. The new version works perfectly with the .NSH script, and i think that we`re past validated your work on newer Intel Models. If you want, you can already update the Repository description to "Tested on Macbook Pro 2019 16'"

Is there a way to contribute to you? I saw a link donation on you profile. Does that help you?

Thanks,
Dan

**Reply by 0nelight** [reply]
Hello @daniel-nextgti,

Great to hear :smiley:
I will add this to the Repository description :sunglasses:

I feel very flattered that you want to donate something to me—certainly every euro/dollar is helpful!

Although I can't judge the authors and their frugality and efficiency in terms of donations—if you look at the two much larger projects, namely memtest86+ and rEFInd, which this project depends on, it would be nice to donate part of it there as well!

Find my donations-page for this project [here](https://www.paypal.com/donate/?hosted_button_id=L4A83Q4J6HFHC). 
Refind has their Donations-Button on the [top of its page](https://www.rodsbooks.com/refind/).
memtest+ has it on the [bottom of its page](https://www.memtest.org/) (altough it seems deactivated right now).

Keep in mind that I will do my best to keep this tool working, but you never know what Apple does in terms of EFI-Firmware Updates over the next years :grimacing: :smiley: 

In case Apple really cares about reducing E-Waste and protecting the Environment - I am sure they will not block this hidden feature of macOs in the future.

Thanks,
0nelight

**Comment by 0nelight** [comment]
[Regarding finding the right paths.](https://github.com/0nelight/macOS-Disable-RAM-Areas/discussions/7#discussioncomment-10735359)

**Comment by adrian77** [comment]
Hi. Good work. It helped me to isolate away 32MB of bad memory from the OS.

There seem to be a way to permanently reserve the memory for firmware instead of re-setting it on every reboot by changing following:
from:
`Status = uefi_call_wrapper(BS->AllocatePages, 4, 2, 8, NumPages, &Addr);`
to
`Status = uefi_call_wrapper(BS->AllocatePages, 4, AllocateAnyPages, EfiRuntimeServicesData, NumPages, &Addr);`

This change tells UEFI to allocate the memory as EfiRuntimeServicesData, a type that can persist after the OS takes control, giving you a better chance of keeping this memory reserved during the macOS boot. However, it’s not guaranteed that macOS will respect this memory reservation unless further configuration is done at the OS or bootloader level.

I will try it and let you know if this works for me (MacBook Pro 16" 2019).

**Comment by adrian77** [comment]
It seems that the change of code did not persist after reboot. Not only macOS did not respect it, but also Memtest86 keep on hitting into bad memory segments.

**Comment by 0nelight** [comment]
Happy this works for you too 😃

AllocatePages is not persistent across reboots. It is possible to write something to NVRAM area, but this doesn't help us regarding telling macOs on every boot: "hey don't touch Ram Area X to Y".
